File: binaryTreeLevelOrderTraversalII.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of lists of integers
    def levelOrderBottom(self, root):
    	if root is None:
    		return []

        currentLevel = [root]
        result = [[root.val]]

        while len(currentLevel) > 0:
            nextLevel = []
            while len(currentLevel) > 0:
            	current = currentLevel[0]
            	currentLevel = currentLevel[1:]
            	if current.left:
            		nextLevel.append(current.left)
            	if current.right:
            		nextLevel.append(current.right)

            if len(nextLevel) > 0:
                result.insert(0, [n.val for n in nextLevel])
                #print [n.val for n in nextLevel]
                currentLevel = nextLevel
        return result

File: JAVA/binaryTreeLevelOrderTraversalII.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<List<Integer>> reverse = new ArrayList<List<Integer>>();
        if (root == null)
            return result;
        ArrayList<TreeNode> cur = new ArrayList<TreeNode>();
        ArrayList<TreeNode> next = new ArrayList<TreeNode>();
        ArrayList<Integer> curLevel = new ArrayList<Integer>();
        cur.add(root);
        while (cur.size()>0){
            
            for (TreeNode n: cur){
                curLevel.add(n.val);
                if (n.left != null)
                    next.add(n.left);
                if (n.right!= null)
                    next.add(n.right);
            }
            cur = next;
            next = new ArrayList<TreeNode>();
            result.add(curLevel);
            curLevel = new ArrayList<Integer>();
        }
        for (int i = result.size()-1; i>=0; i--)
            reverse.add(result.get(i));
        return reverse;
    }
}